File: 2018/python/day1.py
import bisect
import utils


def main():
    lines = utils.readlines("day1/input")
    numbers = [int(l) for l in lines]

    print("Sum of numbers:", sum(numbers))

    currentSum = 0
    i = 0
    knownSums = []
    while True:
        currentSum += numbers[i % len(numbers)]
        pos = bisect.bisect_left(knownSums, currentSum)
        if pos != len(knownSums) and knownSums[pos] == currentSum:
            break
        else:
            bisect.insort(knownSums, currentSum)
            i += 1
    print("First reoccuring sum:", currentSum)


if __name__ == "__main__":
    main()

File: rust/2018/day1/src/day1.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader, Result};
use std::path::Path;

fn main() -> Result<()> {
    let root = env::var("ROOT").unwrap();
    let path = root.to_string() + "/2018/day1/input";
    let file = File::open(&Path::new(&path))?;
    let mut numbers: Vec<i32> = vec![];
    for line in BufReader::new(file).lines() {
        numbers.push(line?.parse().unwrap())
    }

    let mut sum = 0;
    for n in &numbers {
        sum += n
    }
    println!("{}", sum);

    sum = 0;
    let mut i = 0;
    let mut knownsums: Vec<i32> = vec![];
    loop {
        sum += numbers[i % numbers.len()];
        match knownsums.binary_search(&sum) {
            Ok(_) => break,
            Err(pos) => knownsums.insert(pos, sum),
        }
        i += 1;
    }

    println!("{}", sum);
    Ok(())
}
